### ISSUE TITLE
Extracting more standard LogEntry values?

### ISSUE THREAD

**Issue by jum** [nanoandrew4/ngcplogs#7]
I would like to extract more LogEntry values, for example trace ids and labels and such. I do see that client libraries like slogdriver do put them in a special fields in their JSON, where they are interpreted if the app runs in the google cloud environment and the stderr is captured directly. In case of ngcplogs, the output is intercepted indirectly and the values end up in the payload like this:

```
{
  "insertId": "ezyllef2hdhvu",
  "jsonPayload": {
    "instance": {
      "name": "opc4"
    },
    "message": "incoming",
    "container": {
      "imageName": "gitea.mager.org/jum/slogtest:main",
      "imageId": "sha256:540efe2d7c74a09c769cdddda760d4b43bf7300429485661fb17a696b82830af",
      "id": "7b36d4ddffc1d62217bfda728f7699f51b124173e9a6443a7cb71a3a372a9901",
      "created": "2024-05-10T20:23:50.579167976Z",
      "name": "/test"
    },
    "logging.googleapis.com/trace_sampled": true,
    "logging.googleapis.com/trace": "projects/hosting-349701/traces/f4c3a529cae7fefd2144bbe38cff66f7",
    "logging.googleapis.com/spanId": "f66bbf33460c90f6",
    "parent": "00-f4c3a529cae7fefd2144bbe38cff66f7-f66bbf33460c90f6-01",
    "logging.googleapis.com/sourceLocation": {
      "file": "/goapp/main.go",
      "function": "main.main.func3",
      "line": 87
    }
  },
  "resource": {
    "type": "gce_instance",
    "labels": {
      "instance_id": "",
      "zone": "",
      "project_id": "hosting-349701"
    }
  },
  "timestamp": "2024-05-10T23:14:23.030906911Z",
  "severity": "INFO",
  "logName": "projects/hosting-349701/logs/ngcplogs-docker-driver",
  "receiveTimestamp": "2024-05-10T23:14:24.096556056Z"
}
```

I would like to work on getting more of the trace data into my logs. So I believe one has to make the LogEntry that is supposed to be generated available to the extractor function. What do you think? If this is a viable approach?

**Comment by jum** [comment]
Ok, I have found a way to do that. I have now extracted trace information, and caddy web server info. Together I get good logs for the web server and nested as traces other apps that are involved in that request. I can submit pull requests.

**Comment by nanoandrew4** [comment]
Can you provide a clearer example of what is generated now compared to what you want it to look like with these suggested changes?

**Comment by jum** [comment]
I have two sets of changes pending. One is for apps that already format in Google cloud logging format to include labels, trace and source location information. That is useful for example for go apps that use log/slog and [jussi-kalliokoski/slogdriver](https://github.com/jussi-kalliokoski/slogdriver). The second set of changes is for caddy to extract trace and http request information so caddy can be a proper trace parent as per the Google cloud logging conventions. It also helps to make caddy log entries show as proper http requests in the log explorer.

With these changes, and a small middleware function  to parse traceparent headers for slogdriver I can see all my traces for a request grouped under the http request from caddy, helping debug micro services.

**Comment by nanoandrew4** [comment]
Open a PR and lets take a look at your proposal then :) sounds interesting

**Comment by jum** [comment]
I start with extracting label and trace info [pull 1](https://github.com/nanoandrew4/ngcplogs/pull/9)

For simple go apps the following middleware can be used to extract the traceparent header from a http request: [traceparent](https://github.com/jum/traceparent)

**Comment by jum** [comment]
The caddy extract part, depends on the pull above: [pull 2](https://github.com/nanoandrew4/ngcplogs/pull/10)

Caddy needs to to have the tracing directive set for traces to be grouped under the caddy HTTP request.

**Comment by jum** [comment]
Ups, the second PR contains the first one. So I can redo the second standalone after the first is merged if that suits you better.

**Comment by jum** [comment]
Ping: anything new about these pull requests?

**Comment by nanoandrew4** [comment]
Sorry, I have been very busy with work and haven't started up my computer for the last week and a bit. I will review it over the weekend. Thanks for your contributions.

-------- Original Message --------

**Comment by nanoandrew4** [comment]
Issue can be closed since #9 and #11 have been merged. Thank you for your contributions!
